http-body-util/src/combinators/with_trailers.rs

```diff
@@ -0,0 +1,221 @@
+            #[pin]
+            body: T,
+            trailers: Option<F>,
+        },
+        PollTrailers {
+            #[pin]
+            trailers: F,
+            prev_trailers: Option<HeaderMap>,
+        },
+        Done,
+    }
+}
+
+impl<T, F> Body for WithTrailers<T, F>
+where
+    T: Body,
+    F: Future<Output = Option<Result<HeaderMap, T::Error>>>,
+{
+    type Data = T::Data;
+    type Error = T::Error;
+
+    fn poll_frame(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+        loop {
+            let mut this = self.as_mut().project();
+
+            match this.state.as_mut().project() {
+                StateProj::PollBody { body, trailers } => match ready!(body.poll_frame(cx)?) {
+                    Some(frame) => match frame.into_trailers() {
+                        Ok(prev_trailers) => {
+                            let trailers = trailers.take().unwrap();
+                            this.state.set(State::PollTrailers {
+                                trailers,
+                                prev_trailers: Some(prev_trailers),
+                            });
+                        }
+                        Err(frame) => {
+                            return Poll::Ready(Some(Ok(frame)));
+                        }
+                    },
+                    None => {
+                        let trailers = trailers.take().unwrap();
+                        this.state.set(State::PollTrailers {
+                            trailers,
+                            prev_trailers: None,
+                        });
+                    }
+                },
+                StateProj::PollTrailers {
+                    trailers,
+                    prev_trailers,
+                } => {
+                    let trailers = ready!(trailers.poll(cx)?);
+                    match (trailers, prev_trailers.take()) {
+                        (None, None) => return Poll::Ready(None),
+                        (None, Some(trailers)) | (Some(trailers), None) => {
+                            this.state.set(State::Done);
+                            return Poll::Ready(Some(Ok(Frame::trailers(trailers))));
+                        }
+                        (Some(new_trailers), Some(mut prev_trailers)) => {
+                            prev_trailers.extend(new_trailers);
+                            this.state.set(State::Done);
+                            return Poll::Ready(Some(Ok(Frame::trailers(prev_trailers))));
+                        }
+                    }
+                }
+                StateProj::Done => {
+                    return Poll::Ready(None);
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn is_end_stream(&self) -> bool {
+        match &self.state {
+            State::PollBody { body, .. } => body.is_end_stream(),
+            State::PollTrailers { .. } | State::Done => true,
```

Comment:
If I'm reading right, this will mean that while it still has trailers to poll, it will claim the stream is over. So for instance in hyper, the trailers won't be polled.

It probably shouldn't delegate to the inner body either, because it claims to be ended while still in `State::PollBody`, the trailers again won't be polled.

Reply:
Ah yes I see. That makes sense!